### PR TITLE
feat(collector): add Talos OS support via Kubernetes Node info

### PR DESCRIPTION
**Pull request by ayuskauskas**
## Summary

Adds Talos Linux support to the snapshot agent and collector framework, unblocking AICR deployment on clusters that have no systemd and no host `/etc/os-release` accessible to unprivileged pods. Selection is driven by the existing `os` recipe criterion (new `os: talos` value); no new criteria fields are introduced.

## Motivation / Context

Talos has no systemd D-Bus and exposes no `/etc/os-release` at the host filesystem level, so the snapshot agent's privileged pod fails to start: the `/run/systemd` and `/etc/os-release` `hostPath` mounts in `pkg/k8s/agent/job.go` cannot resolve. This is the documented blocker raised in the issue thread.

This PR is the first phase of Talos enablement, scoped to the agent / collector / validator update path. Future phases (declarative cross-OS namespace and security-context configuration; optional Nodewright tuning package for Talos) are deferred to follow-on PRs.

Fixes: N/A
Related: #565

## Type of Change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [ ] Refactoring (no functional changes)
- [ ] Build/CI/tooling

## Component(s) Affected

- [x] CLI (`cmd/aicr`, `pkg/cli`)
- [ ] API server (`cmd/aicrd`, `pkg/api`, `pkg/server`)
- [x] Recipe engine / data (`pkg/recipe`)
- [ ] Bundlers (`pkg/bundler`, `pkg/component/*`)
- [x] Collectors / snapshotter (`pkg/collector`, `pkg/snapshotter`)
- [ ] Validator (`pkg/validator`)
- [x] Core libraries (`pkg/errors`, `pkg/k8s`)
- [x] Docs/examples (`docs/`, `examples/`)
- [ ] Other: ____________

### Selection via existing OS criterion

`recipe.CriteriaOSTalos = "talos"` added to `pkg/recipe/criteria.go` with the standard enum audit (`api/aicr/v1/server.yaml` × 3, `docs/`, `site/docs/`, `pkg/recipe/doc.go`, `pkg/cli/recipe.go`). The collector factory branches on this value via a new `WithOS` option:

```go
func (f *DefaultFactory) CreateSystemDCollector() Collector {
    if f.OS == OSTalos { return talos.NewServiceCollector() }
    return &systemd.Collector{Services: f.SystemDServices}
}

func (f *DefaultFactory) CreateOSCollector() Collector {
    if f.OS == OSTalos { return talos.NewOSCollector() }
    return &os.Collector{}
}
```

`collector.OSTalos` is a string literal kept in sync with `recipe.CriteriaOSTalos` to avoid the `pkg/recipe` → `pkg/snapshotter` → `pkg/collector` import cycle.

### New `pkg/collector/talos` package

Organized so future OS-specific packages can follow the same template:

| File | Role |
|------|------|
| `talos.go` | Package entry: `Option`, `config`, `fetchNode` helper, shared reading keys. |
| `service.go` | `ServiceCollector` emits `TypeSystemD` with `containerd.service` + `kubelet.service` subtypes derived from `Node.Status.NodeInfo.{ContainerRuntimeVersion, KubeletVersion}`. Subtype names match the systemd backend so existing constraints work unchanged. |
| `os.go` | `OSCollector` emits `TypeOS` with a `release` subtype (parsed from `OSImage` / `KernelVersion`) and an `extensions` subtype derived from `extensions.talos.dev/<name>: <version>` Node labels (e.g., `nvidia-container-toolkit`, `nvidia-open-gpu-kernel-modules`). The `extensions` subtype is omitted when no matching labels are present. |
| `doc.go` | Package documentation: layout, selection, escalation path. |

### Pod construction gating

`pkg/k8s/agent/job.go` skips the `/run/systemd` and `/etc/os-release` `hostPath` mounts when `OS == "talos"` and propagates `AICR_OS` to the agent container. The in-pod factory rebuilds itself from `AICR_OS` so it picks the same backend the controller-side picked.

### Deliberate stub: Kubernetes API only

The Talos backend uses only the Kubernetes API surface that `pkg/collector/k8s` already touches. **No MPL-2.0 dependencies are introduced.** The package docs and code comments lay out a future expansion path:

1. **(this PR) Kubernetes API stub** — `Node.Status.NodeInfo` + `Node.Labels`.
2. **gRPC reflection against `machined`** — `google.golang.org/grpc` with `dynamicpb` messages. No vendored MPL surface, but version-fragile and untyped.
3. **Vendor `siderolabs/talos/pkg/machinery/client`** — MPL-2.0 which is different the the rest of dependencies

Step up only when a constraint genuinely needs data the previous phase cannot satisfy.

### CLI

New `--os` flag on `aicr snapshot`, validated through `recipe.ParseCriteriaOSType`. Plumbs through `snapshotter.AgentConfig.OS` → `agent.Config.OS` → pod env + factory.

## Testing

```bash
unset GITLAB_TOKEN

go test -race -short ./...
# 55/55 packages pass

GOFLAGS="-mod=vendor" go test -coverprofile=cover.out ./pkg/collector/talos/...
go tool cover -func=cover.out | tail -1
# total: (statements) 91.7%

golangci-lint run -c .golangci.yaml ./...
# 0 issues.

./tools/check-agents-sync
# OK: AGENTS.md is in sync with .claude/CLAUDE.md
```

Coverage delta: `pkg/collector/talos` is a new package, so this is a baseline (91.7%, well above the 70% project floor). No other package has its coverage decreased by this PR.

New tests added:
- `pkg/collector/talos/service_test.go` — table-driven NodeInfo population, graceful degradation on missing client/node, `splitRuntimeID` edge cases.
- `pkg/collector/talos/os_test.go` — table-driven release + extensions assembly, OSImage parsing across distros (Talos, Ubuntu LTS, RHEL), label-with-empty-suffix filtering, graceful degradation.
- `pkg/collector/factory_test.go` — verifies `OS=talos` returns `*talos.ServiceCollector` / `*talos.OSCollector`; default OS preserves systemd / os backends.
- `pkg/k8s/agent/job_test.go` — `TestBuildPodSpec_TalosSkipsSystemDHostPath` covers Talos / Ubuntu / empty-OS cases for hostPath mounts and `AICR_OS` env.
- `pkg/snapshotter/snapshot_test.go` — `TestParseOSEnv` for the in-pod env-var pickup.
- `pkg/recipe/criteria_test.go` — `talos` in alias and sorted-types tests.

## Risk Assessment

- [x] **Low** — Isolated change, well-tested, easy to revert
- [ ] **Medium** — Touches multiple components or has broader impact
- [ ] **High** — Breaking change, affects critical paths, or complex rollout

**Rollout notes:** Backwards-compatible. All Talos behavior is gated on `OS == "talos"`; existing recipes and agent invocations on Ubuntu / RHEL / COS / AmazonLinux are unchanged. Empty `OS` preserves the systemd default. The new `--os` flag is optional. To revert, drop the feature branch — no migration is required because the wire format adds new constants without changing existing values, and the default factory branches are unchanged.

## Checklist

- [x] Tests pass locally (`make test` with `-race`)
- [x] Linter passes (`make lint`)
- [x] I did not skip/disable tests to make CI green
- [x] I added/updated tests for new functionality
- [x] I updated docs if user-facing behavior changed
- [x] Changes follow existing patterns in the codebase
- [x] Commits are cryptographically signed (`git commit -S`) — [GPG signing info](https://docs.github.com/en/authentication/managing-commit-signature-verification)

### CODE REVIEW

File: pkg/collector/talos/talos.go [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+			slog.String("error", err.Error()))
+		return nil
+	}
+	nodeName := c.resolveNodeName()
+	if nodeName == "" {
+		slog.Warn("node name not set - cannot collect Talos data",
+			slog.String("hint", "agent Job sets NODE_NAME via downward API; check pod env"))
+		return nil
+	}
+	node, err := cs.CoreV1().Nodes().Get(ctx, nodeName, metav1.GetOptions{})
```
Comment (mchmarny) [inline pkg/collector/talos/talos.go:130]: `OSCollector` and `ServiceCollector` each call `fetchNode` independently, so a single Talos snapshot does two `Nodes().Get()` round-trips for the same node — the collectors run in parallel via `errgroup` in `pkg/snapshotter/snapshot.go`. Cheap enough that it isn't blocking, but the natural fix is to fetch the Node once at the snapshotter level (or in a shared `config` cache keyed by node name) and pass it into both build helpers. As more Talos-only collectors get added under this same template the duplication scales linearly.

File: pkg/collector/talos/service.go [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+	if info.ContainerRuntimeVersion == "" {
+		b.SetString(keyActiveState, "unknown")
+		return b.Build()
+	}
+
+	runtimeName, runtimeVersion := splitRuntimeID(info.ContainerRuntimeVersion)
+	b.SetString(keyActiveState, "active").
+		SetString(keyRuntimeName, runtimeName).
+		SetString(keyVersion, runtimeVersion)
```
Comment (mchmarny) [inline pkg/collector/talos/service.go:117]: When `ContainerRuntimeVersion` is non-empty but malformed (no `://`), `splitRuntimeID` returns `(s, "")` and we still stamp `ActiveState=active` with an empty `Version` and the entire raw string as `RuntimeName` — `service_test.go` table case "node with malformed runtime id" locks this in. The empty Version + active state is an inconsistent signal: a downstream constraint comparing `Version >= "1.7"` will silently never match, with no way to tell the data was unparsed.

Two reasonable options:
- Stamp `ActiveState=unknown` when parsing fails (treat malformed runtime ID like missing data), or
- Keep `ActiveState=active` but omit `Version` entirely (don't write the empty string) so constraint engines can distinguish "absent" from "parsed empty".

Either is fine; current behavior locks the worst of both into the test.

File: pkg/collector/factory.go [hunk 3]
```diff
@@ -19,9 +19,16 @@ import (
...
 	"github.com/NVIDIA/aicr/pkg/collector/systemd"
+	"github.com/NVIDIA/aicr/pkg/collector/talos"
 	"github.com/NVIDIA/aicr/pkg/collector/topology"
 )
 
+// OSTalos is the OS criteria value that routes service collection to the
+// Talos backend. Kept as a string literal here (rather than imported from
+// pkg/recipe) to avoid the import cycle pkg/recipe → pkg/snapshotter →
+// pkg/collector. Must stay in sync with recipe.CriteriaOSTalos.
+const OSTalos = "talos"
```
Comment (mchmarny) [inline pkg/collector/factory.go]: `OSTalos` is now defined three places — `recipe.CriteriaOSTalos`, `collector.OSTalos` (here), `agent.OSTalos` — each with a "must stay in sync" comment, but nothing enforces it. A drift between any two would silently route traffic to the wrong backend (e.g., agent strips systemd hostPath but factory still picks systemd because the casing differs).

A one-line guard test in any package that already imports all three (e.g., `pkg/cli` or `pkg/snapshotter`) is enough:

```go
if string(recipe.CriteriaOSTalos) != collector.OSTalos || collector.OSTalos != agent.OSTalos {
    t.Fatal("OSTalos constants drifted; see CLAUDE.md note in factory.go")
}
```

File: pkg/collector/talos/os.go [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+	// parenthesis form (e.g., "Ubuntu 22.04.5 LTS" → "ubuntu").
+	if spaceIdx := strings.Index(id, " "); spaceIdx >= 0 {
+		id = id[:spaceIdx]
+	}
+	if len(parts) == 2 {
+		v := strings.TrimSuffix(strings.TrimSpace(parts[1]), ")")
+		v = strings.TrimPrefix(v, "v")
+		version = v
+	}
+	return id, version
```
Comment (mchmarny) [inline pkg/collector/talos/os.go:160]: `os_test.go` documents that `parseOSImage("Red Hat Enterprise Linux 9.4 (Plow)")` yields `("red", "Plow")` — `id` truncated to `"red"` because of the space-strip, and codename `"Plow"` captured as `VERSION_ID`. This collector is only invoked when `OS == OSTalos` so production data won't hit it, but the test locks in misleading behavior that becomes a trap if a future change reroutes this through here.

Two cleaner options:
- Make the parser Talos-specific: gate on `info.OperatingSystem == "linux"` and the first token being `"talos"`, return empty for anything else. The function is already only meaningful for Talos.
- Keep it permissive but drop the RHEL/Ubuntu test cases and document explicitly that non-Talos OSImage is not supported.

Either way, asserting `("red", "Plow")` as expected output reads as a bug, not a deliberate parse.

File: pkg/snapshotter/snapshot.go [hunk 5]
```diff
@@ -82,13 +82,22 @@ func parseMaxNodesPerEntryEnv() int {
 	return n
 }
 
+// parseOSEnv reads the AICR_OS env var set by the agent Job. Returns the empty
+// string when unset, which preserves the existing default-backend behavior.
+func parseOSEnv() string {
+	return os.Getenv("AICR_OS")
```
Comment (mchmarny) [inline pkg/snapshotter/snapshot.go]: nit: this does no validation. If `AICR_OS=talsoo` (typo) leaks in via the pod env, the in-pod factory's `f.OS == OSTalos` check fails and silently falls back to systemd — which then fails because the systemd hostPath mounts were also skipped on the controller side. Same shape as `parseMaxNodesPerEntryEnv` above, which at least logs `slog.Warn` on invalid input.

Suggest mirroring that pattern: try `recipe.ParseCriteriaOSType(val)`, log `slog.Warn` and return `""` (preserve default) if invalid. Defense-in-depth — the controller-side CLI already validates, but if anything ever sets `AICR_OS` outside that path, a silent fallback is much harder to debug.

File: docs/user/cli-reference.md [hunk 6]
```diff
@@ -316,7 +316,7 @@ Generate recipes using direct system parameters:
 | `--service` | | string | K8s service: eks, gke, aks, oke, kind, lke |
 | `--accelerator` | `--gpu` | string | Accelerator/GPU type: h100, gb200, b200, a100, l40, rtx-pro-6000 |
 | `--intent` | | string | Workload intent: training, inference |
-| `--os` | | string | OS family: ubuntu, rhel, cos, amazonlinux |
+| `--os` | | string | OS family: ubuntu, rhel, cos, amazonlinux, talos |
```
Comment (mchmarny) [inline docs/user/cli-reference.md:320]: The `aicr recipe --os` row is updated to add `talos` here, but the `aicr snapshot` flags table at lines 76–93 still doesn't have an `--os` row at all — that's the brand-new flag this PR introduces in `pkg/cli/snapshot.go`. AGENTS.md is explicit: "When a PR adds or changes user-visible behavior (new CLI flag...), update the relevant page in `docs/` in the same PR — don't defer to a follow-up. CLI flag / subcommand → `docs/user/cli-reference.md`."

Please add a row in the snapshot flags table mirroring the Usage text from the Go flag definition.

**Review by mchmarny — COMMENTED** [review]
Solid first phase of Talos enablement. Scope is appropriately tight (Kubernetes-API stub, no MPL deps, OS gating via existing criterion), the package layout is clearly templated for future OS-specific backends, and test coverage is real (91.7% on the new package, table-driven, graceful-degradation cases included). The three-phase escalation path documented in `pkg/collector/talos/doc.go` is exactly the right framing.

Six findings, none blocking:

- **Doc gap (medium):** the new `aicr snapshot --os` flag isn't in `docs/user/cli-reference.md`'s snapshot flags table — AGENTS.md requires this in the same PR.
- **Duplicate Node Get (medium):** `OSCollector` and `ServiceCollector` each fetch the same Node; cheap today, but scales linearly as more Talos collectors are added under this template.
- **Inconsistent ActiveState semantics (medium):** malformed `ContainerRuntimeVersion` still stamps `ActiveState=active` with empty `Version` — locked in by tests; suggests either reporting `unknown` or omitting the empty Version.
- **Constant drift risk (low/medium):** `OSTalos` is defined in three packages with comments saying "must stay in sync"; trivial guard test would prevent silent drift.
- **`parseOSImage` non-Talos misbehavior (low):** tests document `("Red Hat Enterprise Linux 9.4 (Plow)", "red", "Plow")`. Production never hits this code path (the collector only runs for `os=talos`), but the assertions read like locked-in bugs.
- **`parseOSEnv` silent fallback (low):** invalid `AICR_OS` values silently degrade to systemd default with no log; mirror `parseMaxNodesPerEntryEnv`'s `slog.Warn`.

Three GPU CI checks (nvkind H100/L40G) and `tests/E2E` are still in progress at review time — none of the failures so far are related. CI status looks healthy.

File: pkg/collector/talos/talos.go (hunk 1 above)
Comment (mchmarny) [reply pkg/collector/talos/talos.go:130]: Resolved. `pkg/collector/talos/talos.go` now wraps fetch behind a shared `config` struct with `sync.Once`, and the new `talos.NewCollectors()` constructor returns both collectors against a single shared config. The factory's `ensureTalosCollectors` uses this so a Talos snapshot performs exactly one `Nodes().Get()` round-trip even with the snapshotter's parallel errgroup.

File: pkg/collector/talos/service.go (hunk 2 above)
Comment (mchmarny) [reply pkg/collector/talos/service.go:117]: Resolved. `buildContainerdSubtype` now branches on parse result: non-`containerd` runtime name OR empty version → `ActiveState=unknown` (no `Version` written). The "malformed runtime id" test case is updated to assert this; downstream constraints can now distinguish "absent" from "parsed-but-empty".

File: pkg/collector/factory.go (hunk 3 above)
Comment (mchmarny) [reply pkg/collector/factory.go]: Resolved. New leaf package `pkg/recipe/oskind` is now the single source of truth — `oskind.Talos`, `oskind.IsKnown(...)`, `oskind.All()`. `agent.OSTalos` and `collector.OSTalos` were removed; `recipe`, `collector`, `snapshotter`, and `agent` all import `oskind` directly. No more "must stay in sync" comments and no drift surface.

File: pkg/collector/talos/os.go (hunk 4 above)
Comment (mchmarny) [reply pkg/collector/talos/os.go:160]: Resolved. `parseOSImage` is now Talos-specific: it requires the `Foo (vX)` shape AND `EqualFold(parts[0], "Talos")`, otherwise returns `("", "")` so the caller leaves `ID` / `VERSION_ID` unset rather than emit a misleading parse. The RHEL/Ubuntu test cases are updated to expect empty results.

File: pkg/snapshotter/snapshot.go (hunk 5 above)
Comment (mchmarny) [reply pkg/snapshotter/snapshot.go]: Resolved. `parseOSEnv` now normalizes (`ToLower` + `TrimSpace`), validates against `oskind.IsKnown(val)`, and `slog.Warn`s with the offending value when it's unrecognized — same shape as `parseMaxNodesPerEntryEnv`. `AICR_OS=talsoo` is now logged and falls back to the systemd default explicitly.

File: docs/user/cli-reference.md (hunk 6 above)
Comment (mchmarny) [reply docs/user/cli-reference.md:320]: Resolved. `aicr snapshot` flags table now includes the `--os` row with the full enum and the description of the systemd hostPath / Kubernetes-API backend behavior, mirroring the Go flag's Usage text.

**Review by mchmarny — DISMISSED** [review]
/LGTM 

All 6 of my previous review comments have been resolved

**Review by mchmarny — APPROVED** [review]
